File: flix_backend/lib/flix_backend/data/product.ex
defmodule FlixBackend.Data.Product do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @derive {Phoenix.Param, key: :id}
  @derive {Jason.Encoder,
           only: [
             :id,
             :seller_id,
             :title,
             :description,
             :price,
             :images,
             :category,
             :condition,
             :location,
             :post_time,
             :status,
             :view_count,
             :favorite_count,
             :tags,
             :available_delivery_methods,
             :campus_id
           ]}
  schema "products" do
    # field :seller_id, :string  # 引用 User 的 UID
    field :title, :string
    field :description, :string
    field :price, :float
    field :images, {:array, :string}
    field :category, :string
    field :condition, :string
    field :location, :string
    field :post_time, :integer
    field :status, FlixBackend.Data.ProductStatus, default: :available
    field :view_count, :integer, default: 0
    field :favorite_count, :integer, default: 0

    field :tags, {:array, :string}, default: []
    field :available_delivery_methods, {:array, :string}, default: []

    # 不需要在模型中定义 search_vector 字段，由PostgreSQL在后台管理

    timestamps()

    belongs_to :user, FlixBackend.Data.User,
      foreign_key: :seller_id,
      references: :uid,
      type: :binary_id

    # 添加校区关联
    belongs_to :campus, FlixBackend.Data.Campus,
      foreign_key: :campus_id,
      references: :id,
      type: :binary_id
  end

  def changeset(product, attrs) do
    product
    |> cast(attrs, [
      :seller_id,
      :title,
      :description,
      :price,
      :images,
      :category,
      :condition,
      :location,
      :post_time,
      :status,
      :view_count,
      :favorite_count,
      :tags,
      :available_delivery_methods,
      :campus_id
    ])
    |> validate_required([
      :seller_id,
      :title,
      :description,
      :price,
      :images,
      :category,
      :condition,
      :location
    ])
  end
end

File: Flix/app/src/main/java/fin/phoenix/flix/data/Product.kt
package fin.phoenix.flix.data

import com.google.gson.annotations.SerializedName

/**
 * Product status enum
 * Represents the current status of a product
 */
enum class ProductStatus {
    @SerializedName("available")
    AVAILABLE,     // Available for sale

    @SerializedName("sold")
    SOLD,          // Already sold

    @SerializedName("delete")
    DELETE,        // Deleted by user

    @SerializedName("removed")
    REMOVED        // Removed from listing
}

/**
 * Product data model class
 * Represents a product in the second-hand trading platform
 */
data class Product(
    @SerializedName("id") val id: String, // Product ID, unique identifier (primary key)
    @SerializedName("seller_id") val sellerId: String, // Seller ID, references User table
    @SerializedName("title") val title: String, // Product title
    @SerializedName("description") val description: String, // Product description
    @SerializedName("price") val price: Double, // Product price
    @SerializedName("images") val images: List<String>, // List of product image URLs
    @SerializedName("category") val category: String, // Product category
    @SerializedName("condition") val condition: String, // Product condition (e.g., "New", "90% new")
    @SerializedName("location") val location: String, // Product location/shipping location
    @SerializedName("post_time") val postTime: Long, // Product posting timestamp (milliseconds)
    @SerializedName("status") val status: ProductStatus, // Product status
    @SerializedName("view_count") val viewCount: Int = 0, // Product view count, default 0
    @SerializedName("favorite_count") val favoriteCount: Int = 0, // Product favorite count, default 0
    @SerializedName("tags") val tags: List<String> = emptyList(), // Product tags
    @SerializedName("campus_id") val campusId: String? = null, // Campus ID (可空)
    @SerializedName("available_delivery_methods") val availableDeliveryMethods: List<String> = listOf("express", "pickup") // Available delivery methods
) {
    fun toAbstract(): ProductAbstract {
        return ProductAbstract(
            id = id,
            title = title,
            price = price,
            image = images.firstOrNull() ?: "",
            condition = condition,
            status = status ?: ProductStatus.DELETE,
            campusId = campusId,
                         )
    }
}
